### PR TITLE
AddonVideoCodec: Switch to using CVideoBufferManager

### PR DESCRIPTION
**Pull request by popcornmix**
## Types of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/cores/VideoPlayer/Process/overrides/rbpi/ProcessInfoPi.h [hunk 1]
```diff
@@ -22,13 +22,36 @@
 #include "cores/IPlayer.h"
 #include "../../ProcessInfo.h"
 
+namespace MMAL {
+  class CMMALYUVBuffer;
+};
```
Comment (AlwinEsch) [inline xbmc/cores/VideoPlayer/Process/overrides/rbpi/ProcessInfoPi.h]: Ah `;` is not needed on namespaces

**Comment by popcornmix** [comment]
Ping @FernetMenta @peak3d 
This is a RFC rather than something to merge immediately.
This gets AddonVideoCodec working with MMAL buffers in a platforms independent way.
Other platforms will need similar adjustments.

AddonVideoCodec initially provides just the buffer size desired and the dimensions are only available later which required splitting the Configure API.

Annoyingly it uses the pixel data pointer as an identifier in the API, rather than a general user data pointer, so the map between pixel data pointer and videoBuffer is required.

The whole Configure API and CVideoBufferManager needs more thought when video resolution can change on the fly.

**Comment by FernetMenta** [comment]
> The whole Configure API and CVideoBufferManager needs more thought when video resolution can change on the fly.

@popcornmix a change in resolution triggers a streamchange event in VP. VPV will open the new stream and call CVideoBufferManager::ReleasePools. The will call Released() on the pool. Implementations can use this callback and register a new pool if required.

File: xbmc/cores/VideoPlayer/DVDCodecs/Video/AddonVideoCodec.cpp [hunk 2]
```diff
@@ -330,19 +289,38 @@ void CAddonVideoCodec::Reset()
...
+  m_map.erase(m_lastPictureBuffer);
   m_lastPictureBuffer = nullptr;
 
   m_struct.toAddon.reset(&m_struct);
 }
 
 bool CAddonVideoCodec::GetFrameBuffer(VIDEOCODEC_PICTURE &picture)
 {
-  return m_bufferPool->GetBuffer(picture);
+  CVideoBuffer *videoBuffer = m_videoBufferManager.Get(AV_PIX_FMT_YUV420P, picture.decodedDataSize);
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/AddonVideoCodec.cpp]: what is decodedDataSize and why is this needed?

Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/AddonVideoCodec.cpp]: I see, why don't you drop Get(AVPixelFormat format, int width, int height)? clieant should call Get(fmt, width * height) instead

File: xbmc/cores/VideoPlayer/DVDCodecs/Video/AddonVideoCodec.cpp [hunk 3]
```diff
@@ -263,16 +208,24 @@ CDVDVideoCodec::VCReturn CAddonVideoCodec::GetPicture(VideoPicture* pVideoPictur
...
+    pVideoPicture->iHeight = picture.height;
+    pVideoPicture->pts = picture.pts;
+    pVideoPicture->dts = DVD_NOPTS_VALUE;
+    pVideoPicture->color_range = 0;
+    pVideoPicture->color_matrix = 4;
+    pVideoPicture->iFlags = 0;
+    if (m_codecFlags & DVD_CODEC_CTRL_DROP)
+      pVideoPicture->iFlags |= DVP_FLAG_DROPPED;
+
+    m_videoBufferManager.Configure(AV_PIX_FMT_YUV420P, picture.width, picture.height, picture.stride[0], picture.height, picture.decodedDataSize);
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/AddonVideoCodec.cpp]: clients are not supposed to call internals. clients just request a buffer, the rest is handled internally

Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/AddonVideoCodec.cpp]: see here https://github.com/FernetMenta/kodi-agile/blob/master/xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoPPFFmpeg.cpp#L122

File: xbmc/cores/VideoPlayer/Process/VideoBuffer.cpp [hunk 4]
```diff
@@ -318,14 +318,21 @@ void CVideoBufferPoolSysMem::Return(int id)
   m_free.push_back(id);
 }
 
-void CVideoBufferPoolSysMem::Configure(AVPixelFormat format, int width, int height)
+void CVideoBufferPoolSysMem::Configure(AVPixelFormat format, int width, int height, int aligned_width, int aligned_height, int size)
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/Process/VideoBuffer.cpp]: aligned_width looks like dead code. why is this needed.
btw. we use camelCase here, no "_"

File: xbmc/cores/VideoPlayer/DVDCodecs/Video/AddonVideoCodec.cpp (hunk 2 above)
Comment (popcornmix) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/AddonVideoCodec.cpp]: Will do. 
It's a bit more complicated - YUV has a 3/2 factor,
and you really need to use the aligned width/height.
ffmpeg when decoding H.264 1280x720 will actually require a buffer of 1536x722 due to alignment requirements of various bits of simd code.
But the client that uses the buffer will need to know about that.

File: xbmc/cores/VideoPlayer/Process/VideoBuffer.cpp (hunk 4 above)
Comment (popcornmix) [inline xbmc/cores/VideoPlayer/Process/VideoBuffer.cpp]: aligned width / height are used in the Pi implementation, as that information gets passed back to renderer.
But lets see if I can just use the size on Configure, and work out the aligned_width/aligned_height from GetStride/GetPlanes in the renderer.

Comment (popcornmix) [inline xbmc/cores/VideoPlayer/Process/VideoBuffer.cpp]: Ah - iLineSize/data has gone from VideoPicture, so can't use that for renderer to work out what the aligned width and height are.
So I need the CVideoBuffer (or mmal derived class) to contain this information and some way for AddonVideoCodec to pass that in. Configuring the pool (which the CVideoBuffer has access to) seems like the obvious way.

**Comment by peak3d** [comment]
There should not be a special CVideoBuffer for AddonVideoCodec
AddonVideoCodec should use the existing Buffer(pools) provided by the platform.

File: xbmc/cores/VideoPlayer/Process/VideoBuffer.cpp (hunk 4 above)
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/Process/VideoBuffer.cpp]: ok, then we shoud add stride to Get()

Comment (popcornmix) [inline xbmc/cores/VideoPlayer/Process/VideoBuffer.cpp]: Unfortunately that doesn't work for AddonVideoCodec.
It only provides the buffer size and not the width/height (or stride) when Get is called.
You only see the width/height (and stride) when the decoded picture is returned.

Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/Process/VideoBuffer.cpp]: the client needs a buffer with a set of specific properties. the should be parameters of Get

Comment (popcornmix) [inline xbmc/cores/VideoPlayer/Process/VideoBuffer.cpp]: @peak3d - is it possible to provide width/height/stride at point Get is called (CAddonVideoCodec::GetFrameBuffer)?

Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/Process/VideoBuffer.cpp]: they are parameters of picture (VIDEOCODEC_PICTURE)

Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/Process/VideoBuffer.cpp]: confused ... AddonVideoCodec as a client does not define stride, the buffer pool does

File: xbmc/cores/VideoPlayer/DVDCodecs/Video/AddonVideoCodec.cpp (hunk 3 above)
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/AddonVideoCodec.cpp]: that logic here is wrong btw. at the time the codec outputs a picture is has already requested a buffer and configuration is already defined.

Comment (popcornmix) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/AddonVideoCodec.cpp]: This is the first point in time that picture.stride[0] is filled in.
It is undefined when GetBuffer is called (https://github.com/xbmc/xbmc/pull/12417/commits/ceced4c3f34a4ac04f8392978f0d63e967853c47#diff-7c4d93a31a66a47a05f4bda3208d9edbR310)
@peak3d will have to explain if there is any way round this.
I did initially request getting the width/height/stride when GetBuffer was called but that was not possible with the API widevine provides.

Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/AddonVideoCodec.cpp]: hmm, in general stride is defined by the buffer and the decoder honors this when writing to the buffer rather than the other way round.

**Comment by peak3d** [comment]
libwidevine calls "allocate" callback with only the size needed.
This is done before any picture is produced and therefore there is no way to know currently what will be done with the block of memory.

File: xbmc/cores/VideoPlayer/DVDCodecs/Video/AddonVideoCodec.cpp (hunk 3 above)
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/AddonVideoCodec.cpp]: check out ffmpeg: https://github.com/FFmpeg/FFmpeg/blob/cdd3048134b8b032b69d3c076d73d24e329dc9ba/libavfilter/framepool.c#L51
linesize is set by the pool an honored by decoders. this is why CVideoBuffer provides a method GetStrides and nothing to set those

**Comment by FernetMenta** [comment]
@peak3d as mentioned, stride is defined by the buffer but if this is not possible with widevine we have to add SetStrides to CVideoBuffer

**Comment by peak3d** [comment]
@FernetMenta the decoder is NOT designed by me so I have no influence on it.

**Comment by popcornmix** [comment]
Updated to use "size" as the primary configuration option and SetDimensions as a way of filling in dimensions after allocation.

File: xbmc/cores/VideoPlayer/DVDCodecs/Video/AddonVideoCodec.cpp [hunk 5]
```diff
@@ -263,16 +208,24 @@ CDVDVideoCodec::VCReturn CAddonVideoCodec::GetPicture(VideoPicture* pVideoPictur
...
+    pVideoPicture->iHeight = picture.height;
+    pVideoPicture->pts = picture.pts;
+    pVideoPicture->dts = DVD_NOPTS_VALUE;
+    pVideoPicture->color_range = 0;
+    pVideoPicture->color_matrix = 4;
+    pVideoPicture->iFlags = 0;
+    if (m_codecFlags & DVD_CODEC_CTRL_DROP)
+      pVideoPicture->iFlags |= DVP_FLAG_DROPPED;
+
+    m_videoBufferManager.SetDimensions(picture.width, picture.height, picture.stride[0], picture.height);
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/AddonVideoCodec.cpp]: this should be called on the buffer, not the bufferManager

File: xbmc/cores/VideoPlayer/Process/VideoBuffer.h [hunk 6]
```diff
@@ -161,6 +169,8 @@ class CVideoBufferManager
   void RegisterPool(std::shared_ptr<IVideoBufferPool> pool);
   void ReleasePools();
   CVideoBuffer* Get(AVPixelFormat format, int width, int height);
+  CVideoBuffer* Get(AVPixelFormat format, int size);
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/Process/VideoBuffer.h]: the other Get method is not needed anymore

File: xbmc/cores/VideoPlayer/DVDCodecs/Video/AddonVideoCodec.h [hunk 7]
```diff
@@ -62,5 +62,6 @@ class CAddonVideoCodec
 
   void * m_lastPictureBuffer;
 
-  BufferPool *m_bufferPool;
+  std::map<void *,CVideoBuffer *> m_map;
+  CVideoBufferManager m_videoBufferManager;
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/AddonVideoCodec.h]: a codec is not supposed to own an instance of a buffermanager

File: xbmc/cores/VideoPlayer/DVDCodecs/Video/AddonVideoCodec.cpp [hunk 8]
```diff
@@ -31,70 +31,13 @@ using namespace kodi::addon;
...
-};
-
 CAddonVideoCodec::CAddonVideoCodec(CProcessInfo &processInfo, ADDON::BinaryAddonBasePtr& addonInfo, kodi::addon::IAddonInstance* parentInstance)
   : CDVDVideoCodec(processInfo),
     IAddonInstanceHandler(ADDON_INSTANCE_VIDEOCODEC, addonInfo, parentInstance)
   , m_codecFlags(0)
   , m_displayAspect(0.0f)
   , m_lastPictureBuffer(nullptr)
-  , m_bufferPool(new BufferPool())
+  , m_videoBufferManager(processInfo.GetVideoBufferManager())
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/AddonVideoCodec.cpp]: see above, manager is owned by ProcessIndo

File: xbmc/cores/VideoPlayer/Process/ProcessInfo.h [hunk 9]
```diff
@@ -62,7 +62,7 @@ class CProcessInfo
   bool Supports(EINTERLACEMETHOD method);
   void SetDeinterlacingMethodDefault(EINTERLACEMETHOD method);
   EINTERLACEMETHOD GetDeinterlacingMethodDefault();
-  CVideoBufferManager& GetVideoBufferManager();
+  virtual CVideoBufferManager& GetVideoBufferManager();
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/Process/ProcessInfo.h]: it should not be necessary to override the manager. the manager provices methods for pltforms to register their pools

File: xbmc/cores/VideoPlayer/Process/VideoBuffer.cpp [hunk 10]
```diff
@@ -372,16 +379,30 @@ void CVideoBufferManager::ReleasePools()
 
 CVideoBuffer* CVideoBufferManager::Get(AVPixelFormat format, int width, int height)
 {
+  // todo: fix up other formats
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/Process/VideoBuffer.cpp]: drop this Get method

**Comment by peak3d** [comment]
I think its a good idea to keep both Get() methods. he one with width/height will be the one wich lets Buffer choose the best matching alignement as imlemented today.

**Comment by FernetMenta** [comment]
This would be dead code and keeping dead code is imo not a good idea. It can be implemented again if a client requires it for some reason

**Comment by popcornmix** [comment]
Updated to include passing videoBufferManager to Released callback.
